File: akka-essentials/Chapter7/AkkaSTMExample/src/main/java/org/akka/essentials/stm/java/transactor/example/BankApplication.java
package org.akka.essentials.stm.java.transactor.example;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import org.akka.essentials.stm.java.transactor.example.msg.AccountBalance;
import org.akka.essentials.stm.java.transactor.example.msg.TransferMsg;

public class BankApplication {
    ActorSystem _system = ActorSystem.apply("STM-Example");
    ActorRef bank = _system.actorOf(Props.create(BankActor.class), "BankActor");

    public static void main(String args[]) {

        BankApplication bankApp = new BankApplication();

        bankApp.showBalances();

        bankApp.bank.tell(new TransferMsg(Float.valueOf("1500")), ActorRef.noSender());

        bankApp.showBalances();

        bankApp.bank.tell(new TransferMsg(Float.valueOf("1400")), ActorRef.noSender());

        bankApp.showBalances();

        bankApp.bank.tell(new TransferMsg(Float.valueOf("3500")), ActorRef.noSender());

        bankApp.showBalances();

        bankApp._system.shutdown();

    }

    private void showBalances() {
        try {
            Thread.sleep(2000);
            bank.tell(new AccountBalance("XYZ"), ActorRef.noSender());
            bank.tell(new AccountBalance("ABC"), ActorRef.noSender());

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: akka-essentials/Chapter7/AkkaSTMExample/src/main/scala/org/akka/essentials/stm/transactor/example/BankApplication.scala
package org.akka.essentials.stm.transactor.example
import akka.actor.ActorSystem
import akka.actor.Props
import akka.util.Timeout
import scala.concurrent.duration._
import java.lang.Float

case class AccountBalance(accountNumber: String, accountBalance: Float)
case class AccountCredit(amount: Float)
case class AccountDebit(amount: Float)
case class TransferMsg(amtToBeTransferred: Float)

object BankApplication {
  val system = ActorSystem("STM-Example")
  implicit val timeout = Timeout(5 seconds)
  val bank = system.actorOf(Props[BankActor], name = "BankActor")

  def main(args: Array[String]): Unit = {

    showBalances
    bank ! new TransferMsg(1500f)
    showBalances
    bank ! new TransferMsg(1400f)
    showBalances
    bank ! new TransferMsg(3500f)
    showBalances

    system.shutdown
  }
  def showBalances(): Unit = {
    Thread.sleep(2000)
    bank ! new AccountBalance("XYZ", 0f)
    bank ! new AccountBalance("ABC", 0f)
  }

}